src/lib/y2storage/boot_requirements_strategies/legacy.rb

```diff
@@ -26,48 +26,58 @@ module Y2Storage
   module BootRequirementsStrategies
     # Strategy to calculate the boot requirements in a legacy system (x86 without EFI)
     class Legacy < Base
-      GRUB_SIZE = DiskSize.KiB(256)
-      GRUBENV_SIZE = DiskSize.KiB(1)
-
       def initialize(*args)
         super
         textdomain "storage"
       end
 
       # @see Base#needed_partitions
       def needed_partitions(target)
-        raise Error if grub_in_mbr? && mbr_gap && !valid_mbr_gap?
         planned_partitions = super
         planned_partitions << grub_partition(target) if grub_partition_needed? && grub_partition_missing?
         planned_partitions
       end
 
       # Boot warnings in the current setup
       #
+      # Note that all are just warnings and we don't trigger real errors so far.
+      #
       # @return [Array<SetupError>]
       def warnings
         res = super
 
         if boot_ptable_type?(:gpt)
           res.concat(errors_on_gpt)
-        else
+        elsif boot_ptable_type?(:msdos)
           res.concat(errors_on_msdos)
+        else
+          res.concat(errors_on_plain_disk)
         end
 
         res
       end
 
     protected
 
-      # Whether the MBR gap is big enough
+      # Whether a BIOS GRUB partition will be needed.
       #
-      # @return [Boolean] true if the MBR gap is enough; false otherwise.
-      def valid_mbr_gap?
-        mbr_gap && mbr_gap >= GRUB_SIZE
-      end
-
+      # Note: this is a bit tricky.
+      #
+      # This function is intended to be used in #needed_partitions while
+      # creating a partition proposal. The catch is that when there is no
+      # partition table (yet) it is implicitly assumed that there will be a
+      # gpt created finally - so a grub partition is needed also in this
+      # case.
+      #
+      # FIXME
+      #   It seems that a setup with xen virtual partitions (that are in
+      #   fact disks) also ends up here. In that case no grub partition will
+      #   be needed (below this case is indicated by boot_disk = nil).
+      #   This looks weird.
+      #
+      # @return [Boolean]
       def grub_partition_needed?
-        boot_ptable_type?(:gpt) && grub_part_needed_in_gpt?
+        (boot_ptable_type?(:gpt) || (boot_disk && boot_ptable_type?(nil))) && grub_part_needed_in_gpt?
```

Comment:
This code directly assumes that GPT will be used as default type without asking to `#preferred_ptable_type`. I believe that's wrong. For example, according to fate#323484 is quite likely that we will need to tweak `#preferred_ptable_type` to not always blindly return GPT if running in a Rasberry Pi.

So I would add a flag for `#boot_ptable_type?` to tell it to keep using the old logic (fallback to the preferred type of there is none). Something like:

```ruby
boot_ptable_type?(:gpt, fallback_to_preferred: true) && grub_part_needed_in_gpt?
```

Which, roughly, would be kind of equivalent to:

```ruby
gpt_used =
  if boot_disk && boot_ptable_type?(nil)
    boot_disk.preferred_ptable_type.is?(:gpt)
  else
    boot_ptable_type?(:gpt)
  end
gpt_used && grub_part_needed_in_gpt?
```

Reply:
Ok ok, I agree. I had it changed for this xen virtual partition setup thing anyway. See the reworked patch.

Reply:
:+1: